src/main/scala/is/hail/linalg/BlockMatrix.scala

```diff
@@ -92,12 +92,12 @@ object BlockMatrix {
     new BlockMatrix(blocks, blockSize, nRows, nCols)
   }
 
-  def map4(f: (Double, Double, Double, Double) => Double)(a: M, b: M, c: M, d: M): M =
-    a.map4(b, c, d, f)
-
   def map2(f: (Double, Double) => Double)(l: M, r: M): M =
     l.map2(r, f)
```

Comment:
What do we need map2 for that we can't use mapBlock2?  What about map4?

Reply:
These are both used exclusively in PCRelate. map4 is only used here (the same names are used for the block matrices and their values):
```
    BlockMatrix.map4 { (phi: Double, denom: Double, k2: Double, ibs0: Double) =>
      if (phi <= k0cutoff)
        1.0 - 4.0 * phi + k2
      else
        ibs0 / denom
    } (phi, denom, k2, ibs0)
```
map2 is used in 7 places (5 in the new version), e.g.:
```
    val homalt =
      BlockMatrix.map2 { (g, mu) =>
        if (mu.isNaN || g != 2.0) 0.0 else 1.0
      } (g, mu)
```
PCRelate is also the only user of `map2WithIndex`:
```
    val normalizedGD = g.map2WithIndex(mu, { case (_, i, g, mu) =>
      if (mu.isNaN)
        0.0 // https://github.com/Bioconductor-mirror/GENESIS/blob/release-3.5/R/pcrelate.R#L391
      else {
        val gd = if (g == 0.0) mu
        else if (g == 1.0) 0.0
        else 1.0 - mu

        gd - mu * (1.0 - mu) * twoPhi_ii(i.toInt)
      }
    })
```
`map` and `mapWithIndex` are now unused outside of tests but are simple to implement in terms of the blockMap versions. Not sure whether it's better to delete them or do that.

Either way, in another PR, I think I should re-implementing `map2` and `map2WithIndex` in terms of blockMap versions (I'll need to add blockMap2WithIndex). `map4` is annoying. These were all written at the level of the data underlying the matrices, not clear how much performance will be lost moving the abstraction up, will have to see for PCRelate.

Reply:
Thanks.  It looks like breeze has an indicator function but not a vectorized conditional (numpy.where).  Unclear if using it will be faster than the element-wise stuff.  I saw we revisit this again when we have local ndarrays.